api/hypershift/v1beta1/hostedcluster_types.go

```diff
@@ -1921,34 +1921,152 @@ type AllocationPool struct {
 	End string `json:"end"`
 }
 
-// RouterParam specifies an OpenStack router to use. It may be specified by either ID.
+// RouterParam specifies an OpenStack router to use. It may be specified by either ID or filter, but not both.
 // +kubebuilder:validation:MaxProperties:=1
 // +kubebuilder:validation:MinProperties:=1
 type RouterParam struct {
 	// ID is the ID of the router to use. If ID is provided, the other filters cannot be provided. Must be in UUID format.
 	// +kubebuilder:validation:Format:=uuid
 	// +optional
 	ID *string `json:"id,omitempty"`
+
+	// Filter specifies a filter to select an OpenStack router. If provided, cannot be empty.
+	Filter *RouterFilter `json:"filter,omitempty"`
+}
+
+// RouterFilter specifies a query to select an OpenStack router. At least one property must be set.
+// +kubebuilder:validation:MinProperties:=1
+type RouterFilter struct {
+	Name        string `json:"name,omitempty"`
+	Description string `json:"description,omitempty"`
+	ProjectID   string `json:"projectID,omitempty"`
+
+	FilterByNeutronTags `json:",inline"`
 }
 
-// NetworkParam specifies an OpenStack network. It may be specified by ID.
+func (f *RouterFilter) IsZero() bool {
+	if f == nil {
+		return true
+	}
+	return f.Name == "" &&
+		f.Description == "" &&
+		f.ProjectID == "" &&
+		f.FilterByNeutronTags.IsZero()
+}
```

Comment:
Lets not add these IsZero functions. I added them in CAPO out of an abundance of caution in case somebody upgrading ended up with something like `filter: {}`. We're not upgrading, so we can just not make that mistake in the first place. We can always just test for nil, and additionally use validation to reject a filter with no properties set.

Reply:
done